src/query/plan/preprocess.cpp

```diff
@@ -526,6 +528,18 @@ void Filters::AnalyzeAndStoreFilter(Expression *expr, const SymbolTable &symbol_
   // as `expr1 < n.prop AND n.prop < expr2`.
 }
 
+static void ParseForeach(query::Foreach &foreach, SingleQueryPart &query_part, AstStorage &storage,
+                         SymbolTable &symbol_table) {
+  for (auto *clause : foreach.clauses_) {
+    if (auto *merge = utils::Downcast<query::Merge>(clause)) {
+      query_part.merge_matching.emplace_back(Matching{});
+      AddMatching({merge->pattern_}, nullptr, symbol_table, storage, query_part.merge_matching.back());
```

Comment:
So after deeply understanding what is happening I think you are right, `merge_matching` is handled properly. However, the changes you introduced with `FOREACH`, the above mentioned comment should be modified (and maybe a few other ones around that), because `merge_matching` contains part from clauses that are directly stored in `remaining_clauses`, but inside a clause.